src/graph/sampler.cc

```diff
@@ -766,9 +771,75 @@ DGL_REGISTER_GLOBAL("sampling._CAPI_UniformSampling")
       nflows[i] = new NodeFlow();
-      *nflows[i] = SamplerOp::NeighborUniformSample(
-          gptr, worker_seeds, neigh_type, num_hops, expand_factor, add_self_loop);
+      *nflows[i] = SamplerOp::NeighborSample(
+          gptr, worker_seeds, neigh_type, num_hops, expand_factor,
+          add_self_loop, probability);
+    }
+    return nflows;
+}
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_UniformSampling")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    // arguments
+    const GraphHandle ghdl = args[0];
+    const IdArray seed_nodes = args[1];
+    const int64_t batch_start_id = args[2];
+    const int64_t batch_size = args[3];
+    const int64_t max_num_workers = args[4];
+    const int64_t expand_factor = args[5];
+    const int64_t num_hops = args[6];
+    const std::string neigh_type = args[7];
+    const bool add_self_loop = args[8];
+
+    const GraphInterface *ptr = static_cast<const GraphInterface *>(ghdl);
+    std::vector<NodeFlow*> nflows = _CAPI_NeighborSampling<float>(
+        ptr, seed_nodes, batch_start_id, batch_size, max_num_workers,
+        expand_factor, num_hops, neigh_type, add_self_loop, nullptr);
+
+    *rv = WrapVectorReturn(nflows);
+  });
+
+DGL_REGISTER_GLOBAL("sampling._CAPI_NonUniformSampling")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    // arguments
+    const GraphHandle ghdl = args[0];
+    const IdArray seed_nodes = args[1];
+    const int64_t batch_start_id = args[2];
+    const int64_t batch_size = args[3];
+    const int64_t max_num_workers = args[4];
+    const int64_t expand_factor = args[5];
+    const int64_t num_hops = args[6];
+    const std::string neigh_type = args[7];
+    const bool add_self_loop = args[8];
+    const NDArray probability = args[9];
+
+    const GraphInterface *ptr = static_cast<const GraphInterface *>(ghdl);
+
+    // TODO(BarclayII): template-based selector?
+    std::vector<NodeFlow*> nflows;
+    CHECK(probability->dtype.code == kDLFloat)
+      << "transition probability must be float";
+    CHECK(probability->ndim == 1)
+      << "transition probability must be a 1-dimensional vector";
+    CHECK(probability->shape[0] == ptr->NumEdges())
+      << "transition probability must have same number of elements as edges";
+    CHECK(probability->strides == nullptr || probability->strides[0] == 1)
+      << "transition probability must be contiguous tensor";
+    if (probability->dtype.bits == 32) {
+      const float *prob32 = static_cast<const float *>(probability->data);
+      nflows = _CAPI_NeighborSampling(
+          ptr, seed_nodes, batch_start_id, batch_size, max_num_workers,
+          expand_factor, num_hops, neigh_type, add_self_loop, prob32);
+    } else if (probability->dtype.bits == 64) {
+      const double *prob64 = static_cast<const double *>(probability->data);
+      nflows = _CAPI_NeighborSampling(
+          ptr, seed_nodes, batch_start_id, batch_size, max_num_workers,
+          expand_factor, num_hops, neigh_type, add_self_loop, prob64);
```

Comment:
Looks like you are dispatching implementation based on data types. Check out my implementation in `aten` package (https://github.com/dmlc/dgl/blob/master/src/array/array.cc). You could use a macro. Also, suggest directly use `NDArray` as argument instead of raw `ValueType*` pointer.

Reply:
Which function are you referring to?  I saw in your implementation involving bit-width (like `Full`) you are still dispatching with if-statements.

Reply:
Functions like `Add`, `Sub`.

Reply:
Done